### PR TITLE
Add Function mocking and Object mocking BBE

### PR DESCRIPTION
**Pull request by Ibaqu**
## Purpose
Add Function mocking and Object mocking BBE

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: examples/testerina-function-mocks/testerina_function_mocks.bal [hunk 1]
```diff
@@ -1,29 +1,68 @@
...
+public function addValues(int a, int b) returns int {
+    return intAdd(a, b);
+}
+
+// This function adds two integers and returns the result.
+public function intAdd(int a, int b) returns int {
+    return (a + b);
+}
+
+// The function prints the value of PI using `io:println` function
```
Comment (praneesha) [inline examples/testerina-function-mocks/testerina_function_mocks.bal:14]:
```suggestion
// This function prints the value of PI using the `io:println` function.
```

File: examples/testerina-function-mocks/testerina_function_mocks.bal [hunk 2]
```diff
@@ -1,29 +1,68 @@
...
-    io:println("I'm the mock function!");
-    return (a - b);
+function mockIoPrintLn((any|error)... text) {
+    outputs.push(text);
 }
 
-// This is the test function.
+@test:Config {}
+function testMathConsts() {
+   // This function call to `io:println` will be replaced with `mockIoPrintLn`
```
Comment (praneesha) [inline examples/testerina-function-mocks/testerina_function_mocks.bal:36]:
```suggestion
   // This function call to the `io:println` will be replaced with the `mockIoPrintLn` function.
```

File: examples/testerina-function-mocks/testerina_function_mocks.bal [hunk 3]
```diff
@@ -1,29 +1,68 @@
...
+   printMathConsts();
+   test:assertEquals(outputs[0].toString(), "Value of PI :  3.141592653589793");
+}
+
+// This creates an object for stubbing calls to `intAdd` which is written in the same module.
```
Comment (praneesha) [inline examples/testerina-function-mocks/testerina_function_mocks.bal:41]:
```suggestion
// This creates an object for stubbing calls to the `intAdd` function, which is written in the same module.
```

File: examples/testerina-function-mocks/testerina_function_mocks.bal [hunk 4]
```diff
@@ -1,29 +1,68 @@
...
+@test:MockFn { functionName: "intAdd" }
+test:MockFunction intAddMockFn = new();
+
 @test:Config {}
 function testAssertIntEquals() {
-    int answer = 0;
-    answer = intAdd(5, 3);
-    io:println("Function mocking test");
-    test:assertEquals(answer, 2, msg = "function mocking failed");
+    // This stubs the calls to `intAdd` function to return the specified value.
```
Comment (praneesha) [inline examples/testerina-function-mocks/testerina_function_mocks.bal:47]:
```suggestion
    // This stubs the calls to the `intAdd` function to return the specified value.
```

File: examples/testerina-function-mocks/testerina_function_mocks.bal [hunk 5]
```diff
@@ -1,29 +1,68 @@
...
+    test:when(intAddMockFn).thenReturn(20);
+
+    // This stubs the calls to `intAdd` function to return the specified value
```
Comment (praneesha) [inline examples/testerina-function-mocks/testerina_function_mocks.bal:50]:
```suggestion
    // This stubs the calls to the `intAdd` function to return the specified value
```

File: examples/testerina-function-mocks/testerina_function_mocks.bal [hunk 6]
```diff
@@ -1,29 +1,68 @@
...
+    // when the specified arguments are provided.
+    test:when(intAddMockFn).withArguments(0, 0).thenReturn(-1);
+
+    test:assertEquals(intAdd(10, 6), 20, msg = "function mocking failed");
+    test:assertEquals(intAdd(0, 0), -1,
+        msg = "function mocking with arguments failed");
+
+    // This stubs the calls to `intAdd` function to invoke the specifed function.
```
Comment (praneesha) [inline examples/testerina-function-mocks/testerina_function_mocks.bal:58]:
```suggestion
    // This stubs the calls to the `intAdd` function to invoke the specified function.
```

File: examples/testerina-function-mocks/testerina_function_mocks.bal [hunk 7]
```diff
@@ -1,29 +1,68 @@
...
+    test:when(intAddMockFn).call("mockIntAdd");
+
+    test:assertEquals(addValues(11, 6), 5, msg = "function mocking failed");
 }
 
-// The real function, which is mocked above.
-public function intAdd(int a, int b) returns int {
-    return (a + b);
+// This is a mock function that can be called
+// in place of `intAdd` function.
```
Comment (praneesha) [inline examples/testerina-function-mocks/testerina_function_mocks.bal:65]:
```suggestion
// in place of the `intAdd` function.
```

File: examples/testerina-function-mocks/testerina_function_mocks.description [hunk 8]
```diff
@@ -1,2 +1,6 @@
-// Mock functions allow you to hide the real function and engage your own functions when running tests.
-// This allows you to isolate your test functions from the rest.
+// Mock functions allow you to hide the real functions and engage your own functions when running tests.
```
Comment (praneesha) [inline examples/testerina-function-mocks/testerina_function_mocks.description:1]:
```suggestion
// Mock functions allow you to hide the real functions and engage your own functions when running the tests.
```

File: examples/testerina-function-mocks/testerina_function_mocks.description [hunk 9]
```diff
@@ -1,2 +1,6 @@
...
+// This allows you to isolate your test functions from the rest. Both mocking an imported function and a
+// function declared in the module under test can be mocked with dedicated annotations.<br/><br/>
```
Comment (praneesha) [inline examples/testerina-function-mocks/testerina_function_mocks.description:3]:
```suggestion
// function declared in the module under a test can be mocked with dedicated annotations.<br/><br/>
```

File: examples/testerina-function-mocks/testerina_function_mocks.description [hunk 10]
```diff
@@ -1,2 +1,6 @@
...
+// Mocking an imported function will apply the mocked function to every instance of the original function call.
+// It is not limited to the test file where it is being mocked. Mocking a function from the module under test
```
Comment (praneesha) [inline examples/testerina-function-mocks/testerina_function_mocks.description:5]:
```suggestion
// It is not limited to the test file in which it is being mocked. Mocking a function from the module under the test
```

File: examples/testerina-function-mocks/testerina_function_mocks_test.bal [hunk 11]
```diff
@@ -0,0 +1,69 @@
+// This is the test file that demonstrates ways to mock functions.
```
Comment (praneesha) [inline examples/testerina-function-mocks/testerina_function_mocks_test.bal:1]:
```suggestion
// This is the test file, which demonstrates ways to mock functions.
```

File: examples/testerina-function-mocks/testerina_function_mocks_test.bal [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+public function addValues(int a, int b) returns int {
+    return intAdd(a, b);
+}
+
+// This function adds two integers and returns the result.
+public function intAdd(int a, int b) returns int {
+    return (a + b);
+}
+
+// The function prints the value of PI using `io:println` function
```
Comment (praneesha) [inline examples/testerina-function-mocks/testerina_function_mocks_test.bal:15]:
```suggestion
// This function prints the value of PI using `io:println` function.
```

File: examples/testerina-function-mocks/testerina_function_mocks_test.bal [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+// This demonstrates different ways to mock functions.
+import ballerina/test;
+
+(any|error)[] outputs = [];
+
+@test:Mock {
+    // This specifies a mock function that should replace the
+    // imported function `io:println`.
```
Comment (praneesha) [inline examples/testerina-function-mocks/testerina_function_mocks_test.bal:27]:
```suggestion
    // imported `io:println` function.
```

File: examples/testerina-function-mocks/testerina_function_mocks_test.bal [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+    moduleName: "ballerina/io",
+    functionName: "println"
+}
+function mockIoPrintLn((any|error)... text) {
+    outputs.push(text);
+}
+
+@test:Config {}
+function testMathConsts() {
+   // This function call to `io:println` will be replaced with `mockIoPrintLn`
```
Comment (praneesha) [inline examples/testerina-function-mocks/testerina_function_mocks_test.bal:37]:
```suggestion
   // This function call to the `io:println` function will be replaced with the `mockIoPrintLn` function.
```

File: examples/testerina-function-mocks/testerina_function_mocks_test.bal [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+   printMathConsts();
+   test:assertEquals(outputs[0].toString(), "Value of PI :  3.141592653589793");
+}
+
+// This creates an object for stubbing calls to `intAdd` which is written in the same module.
```
Comment (praneesha) [inline examples/testerina-function-mocks/testerina_function_mocks_test.bal:42]:
```suggestion
// This creates an object for stubbing calls to `intAdd`, which is written in the same module.
```

File: examples/testerina-function-mocks/testerina_function_mocks_test.bal [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+@test:MockFn { functionName: "intAdd" }
+test:MockFunction intAddMockFn = new();
+
+@test:Config {}
+function testAssertIntEquals() {
+    // This stubs the calls to `intAdd` function to return the specified value.
```
Comment (praneesha) [inline examples/testerina-function-mocks/testerina_function_mocks_test.bal:48]:
```suggestion
    // This stubs the calls to the `intAdd` function to return the specified value.
```

File: examples/testerina-function-mocks/testerina_function_mocks_test.bal [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+    test:when(intAddMockFn).thenReturn(20);
+
+    // This stubs the calls to `intAdd` function to return the specified value
```
Comment (praneesha) [inline examples/testerina-function-mocks/testerina_function_mocks_test.bal:51]:
```suggestion
    // This stubs the calls to the `intAdd` function to return the specified value
```

File: examples/testerina-function-mocks/testerina_function_mocks_test.bal [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+    // when the specified arguments are provided.
+    test:when(intAddMockFn).withArguments(0, 0).thenReturn(-1);
+
+    test:assertEquals(intAdd(10, 6), 20, msg = "function mocking failed");
+    test:assertEquals(intAdd(0, 0), -1,
+        msg = "function mocking with arguments failed");
+
+    // This stubs the calls to `intAdd` function to invoke the specifed function.
```
Comment (praneesha) [inline examples/testerina-function-mocks/testerina_function_mocks_test.bal:59]:
```suggestion
    // This stubs the calls to the `intAdd` function to invoke the specified function.
```

File: examples/testerina-function-mocks/testerina_function_mocks_test.bal [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+    test:when(intAddMockFn).call("mockIntAdd");
+
+    test:assertEquals(addValues(11, 6), 5, msg = "function mocking failed");
+}
+
+// This is a mock function that can be called
+// in place of `intAdd` function.
```
Comment (praneesha) [inline examples/testerina-function-mocks/testerina_function_mocks_test.bal:66]:
```suggestion
// in place of the `intAdd` function.
```

File: examples/testerina-function-mocks/testerina_function_mocks_test.out [hunk 20]
```diff
@@ -0,0 +1,20 @@
+# To run this sample, create a Ballerina project and a module inside it.
+# Replace the content of the `main.bal` and ``main_test.bal` files with the above code
```
Comment (praneesha) [inline examples/testerina-function-mocks/testerina_function_mocks_test.out:2]:
```suggestion
# Replace the content of the `main.bal` and ``main_test.bal` files with the above code,
```

File: examples/testerina-object-mocks/testerina_object_mocks.description [hunk 21]
```diff
@@ -0,0 +1,8 @@
+// Object mocking enables controlling the values of member variables and the behavior of
+// the member functions of an object. Mocking of objects can be done in two ways. <br/><br/>
+// 1. Creating a test double - providing an equivalent mock object in place of the real<br/>
+// 2. Stubbing the member function or member variable - stubbing the behavior of functions
+// and values of variables<br/><br/>
```
Comment (praneesha) [inline examples/testerina-object-mocks/testerina_object_mocks.description:5]:
```suggestion
// and values of the variables<br/><br/>
```

File: examples/testerina-object-mocks/testerina_object_mocks_main.bal [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+        http:Request req = new;
+        req.addHeader("Sample-Name", "http-client-connector");
+        result = clientEndpoint->get("/get?test=123", req);
+        response = <http:Response>result;
+        io:println("Status code: " + response.statusCode.toString());
+    }
+    return response;
+}
+
+// This function sends out email to specified email addresses
```
Comment (praneesha) [inline examples/testerina-object-mocks/testerina_object_mocks_main.bal:29]:
```suggestion
// This function sends out emails to the specified email addresses
```

File: examples/testerina-object-mocks/testerina_object_mocks_main.bal [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+// and returns an error if found.
```
Comment (praneesha) [inline examples/testerina-object-mocks/testerina_object_mocks_main.bal:30]:
```suggestion
// and returns an error if they are not found.
```

File: examples/testerina-object-mocks/testerina_object_mocks_test.bal [hunk 24]
```diff
@@ -0,0 +1,89 @@
+// This demonstrates different ways to mock a client object.
+import ballerina/test;
+import ballerina/http;
+import ballerina/email;
+
+// This is the test double of `http:Client` object with the
```
Comment (praneesha) [inline examples/testerina-object-mocks/testerina_object_mocks_test.bal:6]:
```suggestion
// This is the test double of the `http:Client` object with the
```

File: examples/testerina-object-mocks/testerina_object_mocks_test.bal [hunk 25]
```diff
@@ -0,0 +1,89 @@
...
+// implementation of required functions.
```
Comment (praneesha) [inline examples/testerina-object-mocks/testerina_object_mocks_test.bal:7]:
```suggestion
// implementation of the required functions.
```

File: examples/testerina-object-mocks/testerina_object_mocks_test.bal [hunk 26]
```diff
@@ -0,0 +1,89 @@
...
+function testTestDouble() {
+    // This creates and assigns the defined test-double.
+    clientEndpoint = test:mock(http:Client, new MockHttpClient());
+    http:Response res = performGet();
+    test:assertEquals(res.statusCode, 500);
+}
+
+@test:Config {}
+function testReturn() {
+    // This creates and assigns a default mock object which subsequently needs to stubbed.
```
Comment (praneesha) [inline examples/testerina-object-mocks/testerina_object_mocks_test.bal:29]:
```suggestion
    // This creates and assigns a default mock object, which needs to be stubbed  subsequently.
```

File: examples/testerina-object-mocks/testerina_object_mocks_test.bal [hunk 27]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+@test:Config {}
+function testReturnSequence() {
+    http:Response mockResponse = new;
+    mockResponse.statusCode = 404;
+
+    clientEndpoint = test:mock(http:Client);
+    // This stubs the `get` function to return the specified HTTP response
+    // for each call. i.e., The first call will return the status code `200`
```
Comment (praneesha) [inline examples/testerina-object-mocks/testerina_object_mocks_test.bal:61]:
```suggestion
    // for each call (i.e., the first call will return the status code `200`
```

File: examples/testerina-object-mocks/testerina_object_mocks_test.bal [hunk 28]
```diff
@@ -0,0 +1,89 @@
...
+    // and the second call will return the status code `404`.
```
Comment (praneesha) [inline examples/testerina-object-mocks/testerina_object_mocks_test.bal:62]:
```suggestion
    // and the second call will return the status code `404`).
```

File: examples/testerina-object-mocks/testerina_object_mocks_test.bal [hunk 29]
```diff
@@ -0,0 +1,89 @@
...
+    error? err = sendNotification(emailIds);
+    test:assertEquals(err, ());
+
+}
+
+@test:Config {}
+function testMemberVariable() {
+    string mockClientUrl = "http://foo";
+    clientEndpoint = test:mock(http:Client);
+    // This stubs the value of `url` to return the specified string.
```
Comment (praneesha) [inline examples/testerina-object-mocks/testerina_object_mocks_test.bal:85]:
```suggestion
    // This stubs the value of the `url` to return the specified string.
```

File: examples/testerina-function-mocks/testerina_function_mocks.bal (hunk 1 above)
Comment (Ibaqu) [reply examples/testerina-function-mocks/testerina_function_mocks.bal]: Changed as requested

**Pull request merged**